File: test_bigram.cpp
#include <iostream>
#include <cstdio>
#include <string>
#include <fstream>
#include <string>
#include <stack>
#include <vector>
#include <map>
#include <cmath>
#include <boost/algorithm/string.hpp>

#include "typedef.h"
#include "util.h"

using namespace std;


void test(
        const char *modelfile,
        const char *filename,
        const double lambda1=0.90,
        const double lambda2=0.75,
        const long n_unk=1e6,
        const string start_symbol = "<s>",
        const string end_symbol = "</s>"
) {
    // load model
    ifstream modelfs(modelfile);
    string line;

    pair< map<string, double>, map< pair<string, string>, double> > model = load_bimodel(modelfile);
    map<string, double> unimodel = model.first;
    map< pair<string, string>, double> bimodel = model.second;


    ifstream datafs(filename);
    double entropy = 0;
    long total = 0;
    long found = 0;
    while(getline(datafs, line)) {

        vector<string> words, words_without_symbol;
        boost::algorithm::split(words_without_symbol, line, boost::is_space());

        // create words vector
        words.push_back(start_symbol);
        for (auto word : words_without_symbol) {
            words.push_back(word);
        }
        words.push_back(end_symbol);

        for (unsigned long i = 0; i < words.size() - 1; i++) {
            string w0 = words[i];
            string w1 = words[i+1];
            pair<string, string> pr = make_pair(w0, w1);
            double prob = biprob(pr, unimodel, bimodel, lambda1, lambda2, n_unk);

            entropy -= log2(prob);

            // count found word for coverage
            // start_symbol ¤Ï½ü¤¯
            if (i != 0) {
                total++;
                if (unimodel.count(w0)) {
                found += 1;
                }
            }
        }
        // count </s>
        total++;
        found++;
    }
    entropy = entropy / total;
    printf("%ld %ld\n", total, found);
    printf("entropy = %f\n", entropy);
    printf("coverage = %f\n", (double)found / total);

}


int main(int argc, char const* argv[])
{
    const char *modelfile = argv[1];
    const char *filename = argv[2];
    test(modelfile, filename);
    return 0;
}

File: test_bigram.py
#! /usr/bin/env python
# coding:utf-8


import math


def prob(
        words: (str, str),
        unigram: dict,
        bigram: dict,
        lambda1: float,
        lambda2: float,
        u_unk: float
    ):
    if words in bigram:
        pb = lambda2 * bigram[words]
    else:
        w1, w2 = words
        _pb = lambda1 * unigram[w2] if w2 in unigram \
                else (1 - lambda1) * (1 / u_unk)
        pb = (1 - lambda2) * _pb

    return pb


def test(
        modelfile: str,
        filename: str,
        lambda1: float=0.90,
        lambda2: float=0.75,
        n_unk: int=1e6
    ):

    # load model
    unimodel = {}
    bimodel = {}
    with open(modelfile) as f:
        for key, val in (line.rstrip().split("\t") for line in f):
            val = float(val)
            if " " in key:
                bimodel[tuple(key.split())] = val
            else:
                unimodel[key] = val

    entropy = 0
    total = 0
    found = 0

    with open(filename) as f:
        for line in f:
            lst = line.rstrip().split() + ["</s>"]

            w0 = lst[0]
            ent = lambda1 * unimodel[w0] if w0 in unimodel else \
                    (1 - lambda1) * (1 / n_unk)
            entropy += -math.log2(ent)
            for w1, w2 in zip(lst, lst[1:]):
                total += 1
                if w1 in unimodel:
                    found += 1
                else:
                    print(w1, file=open("log1", "a"))
                ent = prob((w1, w2), unimodel, bimodel, lambda1, lambda2, n_unk)
                entropy += -math.log2(ent)
            # count </s>
            total += 1
            found += 1

    entropy = entropy / total
    print(total, found)

    print("entropy = {:.6f}".format(entropy))
    print("coverage = {:.6f}".format(found / total))


if __name__ == '__main__':
    import sys
    modelfile, filename = sys.argv[1:3]
    test(modelfile, filename)
